[WIP] TranslationStore Refactor

benbjohnson opened the pull request:
## Overview

This pull requests refactors the `TranslateStore` to be per-index & per-row and also moves the store implementation from `rhh` to `boltdb`.

_This is still a work in progress but it's already become a large PR so I thought I'd push something to get an initial review & see if anybody had ideas for breaking this up smaller._


## Pull request checklist

- [x] I have read the [contributing guide](https://github.com/pilosa/pilosa/blob/master/CONTRIBUTING.md).
- [x] I have agreed to the [Contributor License Agreement](https://cla-assistant.io/pilosa/pilosa).
- [x] I have updated the [documentation](https://github.com/pilosa/pilosa/tree/master/docs).
- [x] I have resolved any merge conflicts.
- [x] I have included tests that cover my changes.
- [ ] All new and existing tests pass.

## Code review checklist
This is the checklist that the reviewer will follow while reviewing your pull request. You do not need to do anything with this checklist, but be aware of what the reviewer will be looking for.

- [ ] Ensure that any changes to external docs have been included in this pull request.
- [ ] If the changes require that minor/major versions need to be updated, tag the PR appropriately.
- [ ] Ensure the new code is [properly commented](https://github.com/golang/go/wiki/CodeReviewComments#doc-comments) and follows [Idiomatic Go](https://dmitri.shuralyov.com/idiomatic-go).
- [ ] Check that tests have been written and that they cover the new functionality.
- [ ] Run tests and ensure they pass.
- [ ] Build and run the code, performing any applicable integration testing.
- [ ] Make sure PR title conforms to convention in CHANGELOG.md.
- [ ] Make sure PR is tagged with appropriate changelog label.

jaffee commented:
sorry Ben, I've been on vacation but will take a gander as well. don't let
me hold you up.

On Mon, Jul 22, 2019, 11:38 AM Travis Turner <notifications@github.com>
wrote:

> *@travisturner* commented on this pull request.
>
> This seems like the right direction to me. I can't think of a good way to
> break this up into anything smaller.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/pilosa/pilosa/pull/2042?email_source=notifications&email_token=AAHCC4YJMUNUIULDFNHZVTDQAXPB3A5CNFSM4IFPVPLKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB7FERMQ#pullrequestreview-264915122>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHCC4ZQN67FTXNCODUXCVLQAXPB3ANCNFSM4IFPVPLA>
> .
>

benbjohnson commented:
I pushed up some more changes that wires up a lot of the remaining pieces. There's still some outstanding replication test failures that I'm resolving.

benbjohnson commented:
I wanted to drop a note about progress here. The multiplexing of the translation streams has taken longer than expected but it's moving along. I added a replication manager on the `Holder` to manage the replica state (`holderTranslateStoreReplicator`) and I added a multiplexer on the primary side to consolidate all the index & field translation readers into one (`MultiTranslateEntryReader`). 

Since each index and field has their own offset now there's also a `TranslateOffsetMap` which tracks this and allows the remote to request all the streams at the appropriate states in one call. I have unit tests added to test these pieces but I'm working on the integration tests now.

One outstanding issue is what happens when an index or field is created/deleted. My current approach is to just stop replication and refetch using the most recent offset map. This mostly works well (assuming that we aren't creating/deleting indexes & fields constantly). There is some timing issues, however, to ensure that the replica doesn't request from the primary until the new index or field exists.

jaffee commented:
@benbjohnson I've been going through this (sorry I didn't get to it sooner!), and overall, I think it looks great. This approach seems to have simplified things for the most part, and I'm excited about the scalability wins.

One thing we discussed in https://github.com/pilosa/pilosa/issues/2011 that I'm not seeing/understanding is how the sharding of column ID translation is handled. I'm perfectly fine with this being tackled in a future PR, but just wasn't sure if I was missing something. Thanks!

benbjohnson commented:
@jaffee @travisturner I finished refactoring the translate replication and it seems to be working now. There are a few failing tests but they seem to be related to startup and not the translate store specifically. Can you all give this PR a review please? I'll get those few outstanding tests wrapped up.

benbjohnson commented:
I have all the tests passing now with f42b8af. There are, however, some old tests that seem to be flaky and are related to replication timing. I've fixed some of them by adding a `test.RetryUntil()` function but I'm still working through the others as they come up.

I need to rebase now that everything's working but after that it should be good to go. 🚀

jaffee commented:
wooohoo!

jaffee commented:
We should actually talk about how we want to go about incorporating these changes since they necessitate a 2.0 release from a Go API perspective (and from a data format perspective). I think we should merge it to master so this becomes the default tree that we're working on, but maybe add a 2.0-alpha tag or something, so there's a visual indication in the commit history and generated Pilosa version string of when 2.0 started happening? @codysoyland would love your input here (and of course yours as well @benbjohnson).

In some cases (e.g. if translation is not being used), it may be possible to upgrade seamlessly. In others, I think we should at least detect that the translate file is in the wrong place/format, and give a reasonable error (like, "you want 2.0? you can't handle 2.0. please downgrade to 1.x")

I also have had a tiny inkling that we could maybe pretty easily write a small utility to do an offline conversion from the old translatefile style to the new one. The idea being that a lot of the complexity is removed by saying: 
1. turn off Pilosa
2. backup your data directory
3. run this utility on your data directory
4. if it finishes happily, start Pilosa 2.0, if not restore your backup and file an issue.

...rather than needing to manage distributed state and recovery in the case of crashes or whatever, but it's possible that I don't have a good handle on what all could be involved. We definitely don't need to do anything like this before getting this merged.

benbjohnson commented:
@jaffee I have this all working except for one flaky test—`TestClusterTranslator`. The issue is that the translation lives on the primary node now so if it executes  `Set()` against the primary node then it passes but if it queries against the replica then it doesn't replicate back to the primary.

Anyway, I'm adding a redirect now so the replica will send the query to the primary, however, the next step of the translation store is to shard it so I don't want to put too much effort into it since it'll get replaced.

jaffee commented on http/handler.go:
```diff
@@ -306,12 +306,11 @@ func newRouter(handler *Handler) *mux.Router {
 	router.HandleFunc("/internal/nodes", handler.handleGetNodes).Methods("GET").Name("GetNodes")
 	router.HandleFunc("/internal/shards/max", handler.handleGetShardsMax).Methods("GET").Name("GetShardsMax") // TODO: deprecate, but it's being used by the client
-	router.HandleFunc("/internal/translate/data", handler.handleGetTranslateData).Methods("GET").Name("GetTranslateData")
-	router.HandleFunc("/internal/translate/keys", handler.handlePostTranslateKeys).Methods("POST").Name("PostTranslateKeys")
```
oog, I did not notice this change before... I don't suppose there's any way we could continue to support the `/internal/translate/keys` endpoint in a backwards-compatible way? We can update the clients if needed, of course, but it'd be nice to avoid it if possible.

jaffee replied:
with this latest change, I'm now seeing `translate keys: index not found` when I run an ingest which uses this endpoint. (same code works against master)

jaffee replied:
nice, things seem to be working after that last push

benbjohnson commented:
@jaffee I pushed in a `Holder.hasV1TranslateKeysFile()` method and check on it `Open()`. I need to do some integration testing after all those changes I pushed up but tests are finally passing and hopefully this should be nearly ready to merge.

jaffee commented:
yeah, that's working for me, thanks! I'll try to give this another good pass, but I think we can probably get it merged when you're ready.
